examples/xlnet/xlnet/model/decoder.py

```diff
@@ -0,0 +1,194 @@
+            permute_mask = None
+        else:
+            # Only the dummy token is considered target.
+            target_mapping = torch.cat([
+                torch.zeros(1, seq_len - 1, batch_size),
+                torch.ones(1, 1, batch_size)
+            ], dim=1).to(device=word_embed.device)
+            # Dummy token attends to nothing; actual tokens attend to all.
+            permute_mask = torch.cat([
+                torch.zeros(seq_len, seq_len - 1, batch_size),
+                torch.ones(seq_len, 1, batch_size),
+            ], dim=1).to(device=word_embed.device)
+
+        return {
+            "word_embed": word_embed,
+            "segment_ids": segment_ids,
+            "target_mapping": target_mapping,
+            "permute_mask": permute_mask,
+        }
+
+    def initialize(self,  # pylint: disable=no-self-use
+                   helper: Helper, inputs: Optional[Tensor],
+                   sequence_length: Optional[torch.LongTensor],
+                   initial_state: Optional[State]) \
+            -> Tuple[torch.ByteTensor, Tensor, Optional[State]]:
+        initial_finished, initial_inputs = helper.initialize(
+            inputs, sequence_length)
+        return initial_finished, initial_inputs, initial_state
+
+    def step(self, helper: Helper, time: int, inputs: Tensor,
+             state: Optional[State]) \
+            -> Tuple[Output, Optional[State], Tensor, torch.ByteTensor]:
+        self._state_previous_inputs.append(inputs)
+        if not self._state_recompute_memory:
+            assert state is not None
+            net_output, memory = self._forward(
+                memory=state, cache_len=self._state_cache_len, two_stream=True,
+                **self._create_input(self._state_previous_inputs[-1:]))
+            # Omit memory for the dummy token.
+            memory = [mem[:-1] for mem in memory]
+        else:
+            net_output, memory = self._forward(
+                two_stream=True,
+                **self._create_input(
+                    self._state_previous_inputs[-self._state_cache_len:]))
+        logits = F.linear(net_output, self.word_embed.weight, self.lm_bias)
+        logits = logits[-1]
+        sample_ids = helper.sample(time=time, outputs=logits)
+        (finished, next_inputs) = helper.next_inputs(
+            time=time,
+            outputs=logits,
+            sample_ids=sample_ids)
+        outputs = XLNetDecoderOutput(logits=logits, sample_id=sample_ids)
+        return outputs, memory, next_inputs, finished
+
+    def finalize(self, outputs: Output, final_state: Optional[State],
+                 sequence_lengths: torch.LongTensor) \
+            -> Tuple[Output, Optional[State]]:
+        del self._state_cache_len
+        del self._state_recompute_memory
+        del self._state_previous_inputs
+        return super().finalize(outputs, final_state, sequence_lengths)
+
+    @torch.no_grad()
+    def decode(self, start_tokens: torch.LongTensor,
```

Comment:
Do we need a `decode()` method? In `TransformerDecoder`, it's just the `forward()` method, right?

Reply:
Yes -- but I would prefer using separate interfaces for forward (representation computation) vs. decoding. Actually I would prefer separating the interfaces for our existing decoders as well.
The benefits of having separate APIs would be: much cleaner interface (with fewer arguments per method), and cleaner logic inside each method.

Reply:
Currently, our decoders do not have a functionality of  "representation computation", right? I don't think "representation computation" is the main usage of a decoder. Instead, a decoder is for generation, while an encoder is for "representation computation".

If there is any case where users additionally need the representation when generating a sentence, then we may consider adding such functionality. But I doubt the existence of such a use case.

Reply:
Ah yes, you're right, sorry about that. Will fix this.